src/dune_pkg/toolchain.ml

```diff
@@ -278,42 +282,45 @@ let install version =
   Fpath.rm_rf (Path.to_string dest_dir)
 ;;
 
-let get ~log version =
-  let open Fiber.O in
-  let log_print style pp =
-    match log with
-    | `Never -> ()
-    | _ -> User_message.print (User_message.make [ Pp.tag style pp ])
-  in
-  if Version.is_installed version
-  then (
-    (match log with
-     | `Always ->
-       log_print Success
-       @@ Pp.textf
-            "Version %s of the compiler toolchain is already installed in %s"
-            (Version.to_string version)
-            (Version.target_dir version |> Path.Outside_build_dir.to_string)
-     | _ -> ());
-    Fiber.return ())
-  else (
-    let compiler_package = Compiler_package.of_version version in
-    log_print Details
-    @@ Pp.textf
-         "Will install version %s of the compiler toolchain to %s"
-         (Version.to_string version)
-         (Version.target_dir version |> Path.Outside_build_dir.to_string);
-    log_print Details @@ Pp.text "Downloading...";
-    let* () = fetch compiler_package in
-    log_print Details @@ Pp.text "Configuring...";
-    let* () = configure version in
-    log_print Details @@ Pp.text "Building...";
-    let* () = build version in
-    log_print Details @@ Pp.text "Installing...";
-    let+ () = install version in
-    log_print Success
-    @@ Pp.textf
-         "Success! Compiler toolchain version %s installed to %s."
-         (Version.to_string version)
-         (Version.target_dir version |> Path.Outside_build_dir.to_string))
+let get =
+  let mutex = Fiber.Mutex.create () in
```

Comment:
This probably needs an even better mutex to guard against multiple processes trying to create the same compiler. You could take a look at the `git-repo` mutex, which guards against exactly this.

Reply:
Solved by moving the flock used in the rev store to its own file and reusing it here.